src/os/bluestore/BlueStore.cc

```diff
@@ -13598,6 +13704,29 @@ void BlueStore::_choose_write_options(
     }
   );
 
+  wctx->alloc_size = optimal_alloc_size;
+  auto alloc_size_order = optimal_alloc_size_order;
+  if (o->onode.use_small_alloc()) {
+    if (o->onode.size == 0) {
+      o->onode.clear_flag(bluestore_onode_t::FLAG_SMALL_ALLOC);
+    } else {
+      wctx->alloc_size = min_alloc_size;
+      alloc_size_order = min_alloc_size_order;
+    }
+  }
+
+  if (!o->onode.use_small_alloc() &&
+    min_alloc_size != optimal_alloc_size && o->onode.size == 0) {
+    // enable small allocations for tiny object which size is known beforehand
+    if (o->onode.expected_object_size &&
+        o->onode.expected_object_size <= (optimal_alloc_size * 3 / 4)) { //FIXME minor: configure threshold?
+      o->onode.set_flag(bluestore_onode_t::FLAG_SMALL_ALLOC);
+      wctx->alloc_size = min_alloc_size;
+      alloc_size_order = min_alloc_size_order;
+    }
+    // more cases to follow
```

Comment:
it looks like RBD is the only client to set the hint right now, and with its object size isn't likely to benefit from this.

Sage mentioned on-list that RGW sends IMMUTABLE flag hints, but I don't see it doing that - that seems like an easy addition. Given this existing check in bluestore, rgw could set these corresponding flags to prefer larger blobs too: https://github.com/ceph/ceph/blob/8b9ffa696dc30750cd64baaac90590e3c5ef643a/src/os/bluestore/BlueStore.cc#L13610-L13614

Reply:
@jdurgin - yeah, for now this PR is mostly about alloc size selection framework not its usage. The followup is to extend it with set_alloc_hint call for rgw/rados and provide expected_object_size whenever possible. E.g. https://github.com/ceph/ceph/pull/30056/files#diff-c24acfa82f15c38ae9370cab8630a579
performs that. Not sure about setting IMMUTABLE flag, IMO it's redundant at this point as I'd like to use expected_object_size as a trigger  instead.

Reply:
expected_object_size makes sense for the alloc size determination alone, I agree. It seems like IMMUTABLE would be useful for other areas, like compression, dedup, perhaps tiering. Not necessary for this PR though.

Reply:
IMMUTABLE is important for choosing the csum block size...